src/main/cpp/profiler/ProfilerJni.cpp

```diff
@@ -0,0 +1,530 @@
+                "Error enabling overhead activity");
+  } else {
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_DEVICE), "Error enabling device activity");
+    // check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_CONTEXT), "Error enabling context
+    // activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_DRIVER), "Error enabling driver activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_RUNTIME),
+                "Error enabling runtime activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_MEMCPY), "Error enabling memcpy activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_MEMSET), "Error enabling memset activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_NAME), "Error enabling name activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_MARKER), "Error enabling marker activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_CONCURRENT_KERNEL),
+                "Error enabling concurrent kernel activity");
+    check_cupti(cuptiActivityDisable(CUPTI_ACTIVITY_KIND_OVERHEAD),
+                "Error enabling overhead activity");
+    check_cupti(cuptiActivityFlushAll(0), "Error flushing activity records");
+  }
+}
+
+}  // anonymous namespace
+
+}  // namespace spark_rapids_jni::profiler
+
+extern "C" {
+
+using namespace spark_rapids_jni::profiler;
+
+JNIEXPORT void JNICALL Java_com_nvidia_spark_rapids_jni_Profiler_nativeInit(JNIEnv* env,
+                                                                            jclass,
+                                                                            jstring j_lib_path,
+                                                                            jobject j_writer,
+                                                                            jlong write_buffer_size,
+                                                                            jint flush_period_msec)
+{
+  try {
+    setup_nvtx_env(env, j_lib_path);
+    // grab a global reference to the writer instance so it isn't garbage collected
+    auto writer = static_cast<jobject>(env->NewGlobalRef(j_writer));
+    if (!writer) { throw std::runtime_error("Unable to create a global reference to writer"); }
+    State                = new subscriber_state(writer, write_buffer_size);
+    State->writer_thread = std::thread(
+      writer_thread_process, get_jvm(env), writer, write_buffer_size, write_buffer_size);
+    auto rc = cuptiSubscribe(&State->subscriber_handle, callback_handler, nullptr);
+    check_cupti(rc, "Error initializing CUPTI");
+    rc = cuptiEnableCallback(1,
+                             State->subscriber_handle,
+                             CUPTI_CB_DOMAIN_RUNTIME_API,
+                             CUPTI_RUNTIME_TRACE_CBID_cudaDeviceReset_v3020);
+    if (flush_period_msec > 0) {
+      std::cerr << "PROFILER: Flushing activity records every " << flush_period_msec
```

Comment:
nit: I'm not a huge fan of logging to stderr in JNI libraries. Might be good to change this to a logging framework at some point.